OrbitService/ProducerSideServiceImplTest.cpp

```diff
@@ -0,0 +1,244 @@
+class FakeProducer {
+ public:
+  void RunRpc(const std::shared_ptr<grpc::Channel>& channel) {
+    ASSERT_NE(channel, nullptr);
+
+    std::unique_ptr<orbit_grpc_protos::ProducerSideService::Stub> stub =
+        orbit_grpc_protos::ProducerSideService::NewStub(channel);
+    ASSERT_NE(stub, nullptr);
+
+    EXPECT_EQ(context_, nullptr);
+    EXPECT_EQ(stream_, nullptr);
+    context_ = std::make_unique<grpc::ClientContext>();
+    stream_ = stub->ReceiveCommandsAndSendEvents(context_.get());
+    ASSERT_NE(stream_, nullptr);
+
+    read_thread_ = std::thread([this] {
+      orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse response;
+      while (stream_->Read(&response)) {
+        EXPECT_NE(response.command_case(),
+                  orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse::COMMAND_NOT_SET);
+        switch (response.command_case()) {
+          case orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse::kStartCaptureCommand:
+            OnStartCaptureCommandReceived();
+            break;
+          case orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse::kStopCaptureCommand:
+            OnStopCaptureCommandReceived();
+            break;
+          case orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse::COMMAND_NOT_SET:
+            break;
+        }
+      }
+    });
+  }
+
+  void SendBufferedCaptureEvents(int32_t num_to_send) {
+    ASSERT_NE(stream_, nullptr);
+    orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest request;
+    request.mutable_buffered_capture_events();
+    for (int32_t i = 0; i < num_to_send; ++i) {
+      request.mutable_buffered_capture_events()->mutable_capture_events()->Add();
+    }
+    bool written = stream_->Write(request);
+    EXPECT_TRUE(written);
+  }
+
+  void SendAllEventsSent() {
+    ASSERT_NE(stream_, nullptr);
+    orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest request;
+    request.mutable_all_events_sent();
+    bool written = stream_->Write(request);
+    EXPECT_TRUE(written);
+  }
+
+  void FinishRpc() {
+    if (context_ != nullptr) {
+      context_->TryCancel();
+      context_ = nullptr;
+      EXPECT_NE(stream_, nullptr);
+      EXPECT_TRUE(read_thread_.joinable());
+    }
+    stream_ = nullptr;
+    if (read_thread_.joinable()) {
+      read_thread_.join();
+    }
+  }
+
+  MOCK_METHOD(void, OnStartCaptureCommandReceived, (), ());
+  MOCK_METHOD(void, OnStopCaptureCommandReceived, (), ());
+
+ private:
+  std::unique_ptr<grpc::ClientContext> context_;
+  std::unique_ptr<grpc::ClientReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest,
+                                           orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse>>
+      stream_;
+  std::thread read_thread_;
+};
+
+class MockCaptureEventBuffer : public CaptureEventBuffer {
+ public:
+  MOCK_METHOD(void, AddEvent, (orbit_grpc_protos::CaptureEvent && event), (override));
+};
+
+class ProducerSideServiceImplTest : public ::testing::Test {
```

Comment:
Could you also add some tests for the corner cases (errors, no capture data, no AllEventsSend, etc.?).

Reply:
Added some more tests, please have a look.